Add event support for the Catalog Plugin GitLab Discovery Entity providers

elaine-mattos opened the pull request:
## Hey, I just made a Pull Request!

Add events support for the GitLab Entity Provider modules. 
Refs: [RFC: Event Subscription Mechanism for Cross-Plugin Module Integration in the New Backend System #22393](https://github.com/backstage/backstage/issues/22393) and [🚀 Feature: Add Events Support for GitlabDiscoveryEntityProvider and GitlabOrgDiscoveryEntityProvider #21553](https://github.com/backstage/backstage/issues/21553)
#### :heavy_check_mark: TODO List

<!--- Please include the following in your Pull Request when applicable: -->

- [x] Add GitLabOrgDiscoveryEntityProvider module (new Backend)
- [x] Add events support for the GitLabDiscoveryEntityProvider
- [x] Add events support for the GitLabOrgDiscoveryEntityProvider
- [x] Add tests
- [x] Update documentation
- [x] Remove EventSubscriber interface implementation (https://github.com/backstage/backstage/pull/17565#discussion_r1184937825)
- [x] Remove the Events Extension Point and add the [Events Service](https://github.com/backstage/backstage/pull/22344)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

vinzscam commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -240,46 +392,42 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
-        groupUsers = await client.getGroupMembers(group.full_path, ['DIRECT']);
+        groupUsers = await client.getGroupMembers(group.full_path, [
+          'DIRECT',
+          'INHERITED',
```
same comment as `all_available: true` but for `INHERITED` 😄

elaine-mattos replied:
Done, please let me know your thoughts! :)

vinzscam commented on .changeset/dull-sloths-clap.md:
```diff
@@ -0,0 +1,23 @@
+---
+'@backstage/plugin-catalog-backend-module-gitlab-org': minor
```
both look like a `patch` as there will be no breaking change for the adopters, right?

I would suggest splitting this changeset into two separate changesets for each of the packages so that the CHANGELOG for each package will contain only the appropriate changes.

elaine-mattos replied:
Done! :)

vinzscam commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -208,6 +359,7 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
       groups = paginated<GitLabGroup>(options => client.listGroups(options), {
         page: 1,
         per_page: 100,
+        all_available: true,
```
Will `all_available: true` change how the provider behaves for existing adopters? If so, perhaps we could make this change optional through some configuration?

elaine-mattos replied:
Hi @vinzscam ,

so, there is a nice discussion going on about it [here](https://github.com/backstage/backstage/pull/23287). My colleague @markussiebert sent a PR that would allow the parameters for this [api call](https://docs.gitlab.com/ee/api/groups.html#list-groups) to be configurable. However, as @freben said, it would be probably a burden to the majority of the users to have this many options... 

At this time I cannot think of a better solution...I believe having this hardcoded would be benefitial to all users right now as it would allow them to list every group their token has access to, no matter if the token is linked to a regular user or admin. What do you think?

markussiebert replied:
I still think every option that allows filter the results at gitlab api level isn't a bad option ;-)

vinzscam replied:
thank you both, I've missed the discussion. Sorry for causing some confusion 😂
Yeah if `all_available: true` is going to "always be set" then I agree, let's keep it hardcoded.

> I still think every option that allows filter the results at gitlab api level isn't a bad option ;-)

that's absolutely true, I agree! But sometimes it's nice to have less code to maintain, so that we only maintain the configurations that matter.

vinzscam reviewed (changes requested):
amazing work 👏
Just added a tiny comment about the changesets, otherwise ✨

vinzscam commented on plugins/catalog-backend-module-gitlab/src/lib/client.ts:
```diff
@@ -347,6 +383,33 @@ export class GitLabClient {
+    const request = new URL(`${this.config.apiBaseUrl}${endpoint}`);
+    for (const key in options) {
+      if (options[key] !== undefined && options[key] !== '') {
+        request.searchParams.append(key, options[key]!.toString());
```
```suggestion
        request.searchParams.append(key, options[key].toString());
```

elaine-mattos replied:
I would leave this one as it is because we're already checking if the options exist in line 393 ( and line 358, where the same construction was used before). If I remove the non-null assertion operator, even if the check has been done before, Typescript warns me that the "Object is possibly 'undefined'". 
Is there any other way around this?

vinzscam replied:
ah good point I think something like this should work:

```js
    for (const key in options) {
      const value = options[key];
      if (value !== undefined && value !== '') {
        request.searchParams.append(key, value.toString());
      }
    }
```

so that we are more protected in case the `if` gets removed in the future

elaine-mattos replied:
Done!

vinzscam commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -321,6 +478,304 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+      });
+    }
+
+    if (usersToBeAdded.length !== 0) {
```
if we receive a `user_remove_from_group` event and `getGroupMembers()` returns empty array we don't modify spec.members. Is it correct?

elaine-mattos replied:
Hi, yes, this is correct!

elaine-mattos replied:
Hi, 
yeah, you're right. Again! :)
Is it better now? 

```typescript
    // update members of the group entity
    groupEntityToModify[0].spec.members =
      usersToBeAdded.length !== 0 ? usersToBeAdded.map(e => e.username) : [];
```

vinzscam replied:
yeah sounds good! You could also skip the check `groupEntityToModify[0].spec.members = usersToBeAdded.map(e => e.username);` but it's ok!

vinzscam commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -321,6 +478,304 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+      return;
+    }
+
+    // get members of the group
+    const usersToBeAdded: string[] = [];
+    let groupMembers: PagedResponse<GitLabUser> = { items: [] };
+
+    const relations: string[] =
+      this.config.allowInherited === true && this.config.allowInherited
+        ? ['DIRECT', 'INHERITED']
+        : ['DIRECT'];
+
+    groupMembers = await client.getGroupMembers(
+      groupToRebuild.full_path,
+      relations,
+    );
+
+    if (groupMembers.items.length !== 0) {
+      groupMembers.items.forEach(element => {
+        if (
+          event.event_name === 'user_remove_from_group' &&
+          element.username === event.user_username
+        ) {
+          return;
+        }
+
+        usersToBeAdded.push(element.username);
+      });
+    }
```
if I've read the logic correctly this could be simplified a bit:

```suggestion
    const relations = this.config.allowInherited
      ? ['DIRECT', 'INHERITED']
      : ['DIRECT'];

    const groupMembers = await client.getGroupMembers(
      groupToRebuild.full_path,
      relations,
    );

    // new members of the group
    const usersToBeAdded =
      event.event_name === 'user_remove_from_group'
        ? groupMembers.items.filter(e => e.username !== event.user_username)
        : groupMembers.items;
    }
```

then below
```js
usersToBeAdded.map(e => e.username)
```

elaine-mattos replied:
Great remark, thanks! :)

vinzscam commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -240,46 +392,44 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
       let groupUsers: PagedResponse<GitLabUser> = { items: [] };
       try {
-        groupUsers = await client.getGroupMembers(group.full_path, ['DIRECT']);
+        const relations: string[] =
+          this.config.allowInherited === true && this.config.allowInherited
+            ? ['DIRECT', 'INHERITED']
+            : ['DIRECT'];
```
```suggestion
        const relations = this.config.allowInherited
          ? ['DIRECT', 'INHERITED']
          : ['DIRECT'];
```

elaine-mattos replied:
Done

elaine-mattos commented:
Hi @vinzscam !

I had some issues with my branch and had to recreate it (I know, I'm such a git newbie! :D ).
I did incorporate your suggestions/remarks. Could you please take a look at them?
Also, the pipeline is throwing a formatting error on a file I did not touch (beps/0004-scaffolder-task-idempotency/README.md). Should I do anything about it?

thanks again in advance!!! :)

elaine-mattos commented on .changeset/eleven-pandas-divide.md:
```diff
@@ -0,0 +1,7 @@
+
+- Added events support
+- Refactored tests
+- Added documentation
```
Done!

elaine-mattos commented on .changeset/shy-students-clap.md:
```diff
@@ -0,0 +1,10 @@
+- Refactored tests
+- Added the `allowInherited` configuration option that allows groups to be ingested with inherited/indirect member as well.
+- Added `all_available` option (still hardcoded - always true) to the `listGroups` call. PS.: A better solution is on its way, see: <https://github.com/backstage/backstage/pull/23287>
+- Added documentation
```
Also done!

freben commented:
Yeah that `prettier` error has been fixed since then. Rebasing on a fresh master gets rid of the problem. Something like this

```
git checkout master
git pull
git checkout feat/add-event-support-gitlab-provider
git pull
git rebase master
```

elaine-mattos commented:
Hi guys,

the suggested changes have been made. Could you please take a look?
I appreciate the feedback :)

pjungermann commented on plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -0,0 +1,55 @@
+  coreServices,
+  createBackendModule,
+} from '@backstage/backend-plugin-api';
+import { GitlabOrgDiscoveryEntityProvider } from '@backstage/plugin-catalog-backend-module-gitlab';
```
_a question for the maintainers; also rather for a follow-up_

I know that this follows the same pattern as used for the GitHub org provider. However, I have the same question there.

As we create a separate module and package for `gitlab-org`, wouldn't it be better to move the provider to this package and deprecate it at the other one (e.g., by importing it from this new package and reexporting it, marked as deprecated -- similar to few other places).

What do you think?

elaine-mattos replied:
Hi @pjungermann ,

 I would be glad to do it if provided with an example! :)

pjungermann replied:
@vinzscam yeah, nothing that blocks this PR.

I'm also not sure what the maintainers opinion is on that one. I think it would make sense to split that up.
It does affect other plugins like the one for github-org as well.

vinzscam replied:
I think the issue is that `gitlab` can't import from `gitlab-org`, as this will cause a circular dependency

elaine-mattos commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -321,6 +475,288 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+    // new members of the group
+    const usersToBeAdded =
+      event.event_name === 'user_remove_from_group'
+        ? groupMembers.items.filter(e => e.username !== event.user_username)
```
You're right! I have changed it to `const usersToBeAdded = groupMembers.items;`

elaine-mattos commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:
```diff
@@ -321,6 +475,288 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+        web_url: '',
+        avatar_url: '',
+      };
+    } else {
```
Done! :)

elaine-mattos commented:
Hi!
I have incorporated the changes requested by @vinzscam .  However, I'm trying to update the feature branch but I'm getting tons of vale errors. My vale version is the 3.3.1. 
I can fix the errors by adding backticks around the unrecognized words in the changelog.md files. The question is, should I? :)

elaine-mattos commented:
Hi @vinzscam ,

great news! 😄 
We caught a small bug in our environment and fixed it. The `allowInherited` config option was not being correctly read from the config file because it was not added to the `readGitlabConfig` function (my bad 🥲 ). I integrated the fix in the last commit and also rebased to get rid of the merge conflict.

awanlin reviewed (changes requested):
Hi @elaine-mattos, can you please add details about the fact that you need to define the schedule in your provider config when using the new backend system? This is missing in the current docs and I don't see it here. Thanks!

elaine-mattos commented:
> Hi @elaine-mattos, can you please add details about the fact that you need to define the schedule in your provider config when using the new backend system? This is missing in the current docs and I don't see it here. Thanks!

Hi @awanlin ! I've changed the docu a little bit. Could you please review? :)

awanlin commented on docs/integrations/gitlab/org.md:
```diff
@@ -66,7 +168,13 @@ catalog:
         group: org/teams # Required for gitlab.com when `orgEnabled: true`. Optional for self managed. Must not end with slash. Accepts only groups under the provided path (which will be stripped)
+        allowInherited: true # Allow groups to be ingested even if there are no direct members.
         groupPattern: '[\s\S]*' # Optional. Filters found groups based on provided pattern. Defaults to `[\s\S]*`, which means to not filter anything
+        schedule: # optional; same options as in TaskScheduleDefinition
```
Is this true in all cases that it's optional? If I'm using the New Backend System would that not then be required?

elaine-mattos replied:
Hi @awanlin !

I changed this line. Could you please check? :)

awanlin commented on docs/integrations/gitlab/discovery.md:
```diff
@@ -59,19 +120,45 @@ builder.addEntityProvider(
+        group: example-group # Optional. Group and subgroup (if needed) to look for repositories. If not present the whole instance will be scanned
+        entityFilename: catalog-info.yaml # Optional. Defaults to `catalog-info.yaml`
+        projectPattern: '[\s\S]*' # Optional. Filters found projects based on provided patter. Defaults to `[\s\S]*`, which means to not filter anything
+        schedule: # optional; same options as in TaskScheduleDefinition
```
Is this true in all cases that it's optional? If I'm using the New Backend System would that not then be required?

elaine-mattos replied:
Changed this line as well!

awanlin reviewed (commented):
Thanks for adding in the schedule details to the docs @elaine-mattos, I left two comments but they are the same. As I haven't reviewed the code just the docs I'm only leaving this review as a comment

elaine-mattos commented:
Hi @vinzscam, @freben , @awanlin ,

is there anything else I can do to get this one merged? 
Please let me know if there is anything missing! :)

freben commented on docs/integrations/gitlab/org.md:
````diff
@@ -11,34 +11,134 @@ groups -- directly from GitLab. The result is a hierarchy of
 ```bash
 # From your Backstage root directory
-yarn --cwd packages/backend add @backstage/plugin-catalog-backend-module-gitlab
+yarn --cwd packages/backend add @backstage/plugin-catalog-backend-module-gitlab @backstage/plugin-catalog-backend-module-gitlab-org
````
really both?

freben commented on plugins/catalog-backend-module-gitlab/package.json:
```diff
@@ -1,72 +1,76 @@
+    "@backstage/plugin-catalog-common": "workspace:^",
     "@backstage/plugin-catalog-node": "workspace:^",
+    "@backstage/plugin-events-node": "workspace:^",
+    "@gitbeaker/rest": "^40.0.3",
```
```suggestion
    "@gitbeaker/rest": "^39.25.0",
```

if that's okay, to align with the other deps in the repo - this should clean up the yarn.lock diff

freben commented on plugins/catalog-backend-module-gitlab/package.json:
```diff
@@ -1,72 +1,76 @@
+      ]
+    }
+  },
+  "version": "0.3.15-next.1"
```
not sure what happened with the field order sorting here :) version is usually way up at the top etc, but we can sort (hah!) that out some other time

freben commented on plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts:
```diff
@@ -33,26 +36,38 @@ import {
+ *
  * @public
  */
+// <<< EventSupportChange: implemented EventSubscriber interface
```
remove

freben reviewed (approved):
OK, phew! I left some comments that can be deferred to a follow-up PR some time down the line. Do check them out. But for now, :shipit:
